LibraBFT/ImplShared/Util/Crypto.agda

```diff
@@ -29,18 +29,38 @@ module LibraBFT.ImplShared.Util.Crypto where
 
   open WithCryptoHash sha256 sha256-cr
 
-  -- IMPL-DIFF
-  -- 1. cannot do "qcParts" because bs-concat does returns BitString
-  -- 2. no need to deal with a 'Proposal' for stable test hashes
-  hashBD : BlockData → HashValue
-  hashBD = sha256 ∘ bs-concat ∘ blockDataBSList
-   where
-    blockDataBSList : BlockData → List ByteString
-    blockDataBSList (BlockData∙new
-                      epoch round
-                      (QuorumCert∙new voteData (LedgerInfoWithSignatures∙new ledgerInfo _sigs))
-                      blockType) =
-      encode epoch ∷ encode round ∷ encode voteData ∷ encode ledgerInfo ∷ encode blockType ∷ []
+  -- We do not yet have sufficient support to model precisely the hashing used in the Haskell code,
+  -- so it is better that we postulate its intended properties to ensure proofs have the desired
+  -- properties available.  Note that a simpler injectivity property such as
+  --
+  -- hashBD-inj' : ∀ {bd1} {bd2}
+  --             → hashBD bd1 ≡ hashBD bd2
+  --             → NonInjective-≡ sha256
+  --             ⊎ bd1 ≡L bd2
+  --
+  -- does *hot* hold because the input to the hash function used in hashBD does *not* include all
+  -- components of the BlockData being hashed.
+
+  -- TODO-2: Similarly express other hashing functions and their intended injectivity properties where
+  -- the Agda implementation does not accurately reflect the Haskell implementation.
+
+  -- TODO-2: Enable support for the hashS function we use in the Haskell code, which hashes tuples
+  -- of serializable types), and then define hashBD using this and prove that it ensures hashBD-inj.
+  -- Note also the TODO below related to HashTags.
+  postulate
+    hashBD     : BlockData → HashValue
+    hashBD-inj : ∀ {bd1 bd2}
```

Comment:
This has a rather unwieldy return type. If you're ok with this, I would like to package all of these pieces together into a record (happy to take name suggestions!)

Reply:
Good suggestion @cwjnkins, thanks.  I have offered a name and padded it out with a definition :smile:.